src/languages/internal/languagesservice.cpp

```diff
@@ -53,470 +59,340 @@ static QString downloadingStatusTitle()
-    QJsonDocument jsonDoc = QJsonDocument::fromJson(json, &err);
-    if (err.error != QJsonParseError::NoError || !jsonDoc.isObject()) {
-        LOGE() << "failed parse, err: " << err.errorString();
-        result.ret = make_ret(Err::ErrorParseConfig);
-        return result;
+    for (QTranslator* t : m_translators) {
+        qApp->removeTranslator(t);
+        delete t;
     }
+    m_translators.clear();
 
-    result.ret = make_ret(Err::NoError);
-    QStringList languages = jsonDoc.object().keys();
-    for (const QString& key : languages) {
-        if (!jsonDoc.object().value(key).isObject()) {
-            continue;
+    for (const io::path_t& file : lang.files) {
+        QTranslator* translator = new QTranslator();
+        bool ok = translator->load(file.toQString());
+        if (ok) {
+            qApp->installTranslator(translator);
+            m_translators << translator;
+        } else {
+            LOGE() << "Error loading translator " << file.toQString();
+            delete translator;
         }
-
-        QJsonObject value = jsonDoc.object().value(key).toObject();
-
-        Language language;
-        language.code = key;
-        language.name = value.value("name").toString();
-        language.archiveFileName = value.value("file_name").toString();
-        language.files = parseLanguageFiles(value);
-
-        result.val.insert(key, language);
     }
 
-    return result;
-}
-
-LanguageFiles LanguagesService::parseLanguageFiles(const QJsonObject& languageObject) const
-{
-    LanguageFiles files;
+    QLocale locale(lang.code);
+    QLocale::setDefault(locale);
+    qApp->setLayoutDirection(locale.textDirection());
 
-    for (const QString& fileType: languageFileTypes) {
-        QJsonObject instrumentsFile = languageObject.value(fileType).toObject();
-        if (!instrumentsFile.isEmpty()) {
-            files << LanguageFile(instrumentsFile.value("file_name").toString(),
-                                  instrumentsFile.value("hash").toString());
+    if (m_inited) {
+        if (QQmlEngine* engine = uiEngine()->qmlEngine()) {
+            engine->retranslate();
         }
     }
 
-    return files;
+    m_currentLanguage.set(lang);
 }
 
-bool LanguagesService::isLanguageExists(const QString& languageCode) const
+QString LanguagesService::effectiveLanguageCode(const QString& languageCode)
 {
-    io::paths_t files = configuration()->languageFilePaths(languageCode);
-    return !files.empty();
-}
+    auto hasCode = [this](const QString& code) -> bool {
+        return m_languagesHash.contains(code);
+    };
 
-bool LanguagesService::checkLanguageFilesHash(const QString& languageCode, const LanguageFiles& languageFiles) const
-{
-    io::paths_t filePaths = configuration()->languageFilePaths(languageCode);
-    int filesSize = static_cast<int>(filePaths.size());
-    if (filesSize != languageFiles.size()) {
-        return false;
-    }
+    auto tryCode = [hasCode](QString code) -> QString {
```

Comment:
What does this lambda do?

Reply:
It tries decreasingly specific versions of the specified language code. For example, the system ui language might be `nl_NL`. This lambda tries if we have that. That's not the case, so we try a simpler version, just `nl`. We do have that one, so it will be returned by this lambda.